File: src/services/billing_notification_service.py
from sqlalchemy.orm import Session
from src.models.billing import Subscription
from src.services.invoicing_service import InvoicingService
from datetime import datetime, timedelta
from loguru import logger

class BillingNotificationService:
    """
    Controlador de Retenção e Alertas de Faturamento.
    Replica o 'SubscriptionGuard' do .NET.
    """
    @staticmethod
    async def process_billing_heartbeat(db: Session):
        """
        Varredura periódica para gerenciar o ciclo de vida das assinaturas.
        Executado via background task no main.py.
        """
        now = datetime.utcnow()
        
        # 🟢 1. Alertas de Pré-Vencimento (3 dias antes)
        three_days_warning = now + timedelta(days=3)
        subs_to_warn = db.query(Subscription).filter(
            Subscription.expires_at <= three_days_warning,
            Subscription.expires_at > now,
            Subscription.status == "active"
        ).all()
        
        for sub in subs_to_warn:
            # TODO: Enviar WhatsApp/Email real via NotificationService
            logger.info(f"⏰ Alerta de Vencimento enviado para Tenant {sub.tenant_id} (Expira em: {sub.expires_at})")

        # 🟢 2. Suspensão Automática (Expirados)
        expired_subs = db.query(Subscription).filter(
            Subscription.expires_at < now,
            Subscription.status == "active"
        ).all()
        
        for sub in expired_subs:
            sub.status = "past_due"
            logger.warning(f"🚫 Tenant {sub.tenant_id} suspenso por expiração de plano.")
            
            # Gera a fatura do novo ciclo para o cliente pagar e reativar
            InvoicingService.generate_monthly_invoice(db, sub.tenant_id)
            
        db.commit()

File: node-version/src/services/billing/billingNotificationService.js
const { Subscription } = require('../../models/sql/models');
const InvoicingService = require('./invoicingService');
const logger = require('../../utils/logger');
const moment = require('moment');
const { Op } = require('sequelize');

class BillingNotificationService {
  
  static async processBillingHeartbeat() {
    const now = moment();

    // 1. Alerta de Vencimento
    const threeDaysWarning = moment().add(3, 'days');
    
    // Ignora Tenancy pois é um Cron global
    const subsToWarn = await Subscription.findAll({
      where: {
        expires_at: {
          [Op.lte]: threeDaysWarning.toDate(),
          [Op.gt]: now.toDate()
        },
        status: 'active'
      },
      ignoreTenant: true
    });

    for (const sub of subsToWarn) {
      logger.info(`⏰ Alerta de Vencimento enviado para Tenant ${sub.tenant_id} (Expira em: ${sub.expires_at})`);
    }

    // 2. Suspensão Automática
    const expiredSubs = await Subscription.findAll({
      where: {
        expires_at: { [Op.lt]: now.toDate() },
        status: 'active'
      },
      ignoreTenant: true
    });

    for (const sub of expiredSubs) {
      sub.status = 'past_due';
      await sub.save({ ignoreTenant: true });
      logger.warn(`🚫 Tenant ${sub.tenant_id} suspenso por expiração de plano.`);
      
      try {
        await InvoicingService.generateMonthlyInvoice(sub.tenant_id);
      } catch(e) {
        logger.error(`Falha ao gerar fatura pós suspensão: ${e.message}`);
      }
    }
  }
}

module.exports = BillingNotificationService;
